// @beruang/brown/brown-behavior.js
import {dedupingMixin} from '@polymer/polymer/lib/utils/mixin.js';

let internalBrownBehavior = (base) =>
class extends base {
	static get properties() {
		return {
			topLeft: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			topCenter: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			topRight: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			midLeft: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			midCenter: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			midRight: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			bottomLeft: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			bottomCenter: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			},

			bottomRight: {
				type: Boolean,
				value: false,
				reflectToAttribute: true
			}
    	};
	}
}

export const BrownBehavior = dedupingMixin(internalBrownBehavior);
